File: tests/DesktopGrass.Win2D.Tests/SimTests/BirdFlybyTests.cs
// BirdFlybyTests.cs - §17.8 ambient bird flyby tests.

using System;
using System.Collections.Generic;
using System.Linq;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class BirdFlybyTests
{
    private const double Monitor1920 = 1920.0;

    private static Sim BuildSim(ulong seed = Constants.CANONICAL_TEST_SEED)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(seed, Monitor1920, Constants.DEFAULT_DENSITY),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            CurrentScene = Scene.Grass,
        };
        sim.ResetAmbientGusts(seed, Monitor1920);
        sim.ResetEntities(seed);
        sim.Entities.Clear();
        return sim;
    }

    private static int CountBirds(Sim sim) => sim.Entities.Count(e => e.Kind == EntityKind.Bird);
    private static List<Entity> Birds(Sim sim) => sim.Entities.Where(e => e.Kind == EntityKind.Bird).ToList();

    private static int PrngCount(ref Prng side, int minCount, int maxCount)
    {
        double draw = side.Uniform(minCount, maxCount + 1);
        int count = (int)Math.Floor(draw);
        if (count < minCount) count = minCount;
        if (count > maxCount) count = maxCount;
        return count;
    }

    private static void ResetBirdStreamFresh(Sim sim, ulong seed)
    {
        sim.BirdFlybyPrng = Prng.Init(seed ^ Constants.BIRD_FLYBY_PRNG_SALT);
        sim.NextBirdFlybyAtTime = sim.GlobalTime;
    }

    private static void ResetBirdSchedule(Sim sim, ulong seed)
    {
        sim.BirdFlybyPrng = Prng.Init(seed ^ Constants.BIRD_FLYBY_PRNG_SALT);
        sim.NextBirdFlybyAtTime = sim.GlobalTime + Sim.BirdFlybySampleInterval(ref sim.BirdFlybyPrng);
    }

    private static ulong FindVSeed(int minSize)
    {
        for (ulong i = 1; i < 10000; i++)
        {
            ulong seed = Constants.CANONICAL_TEST_SEED + i;
            var sim = BuildSim(seed);
            ResetBirdStreamFresh(sim, seed);
            sim.SpawnBirdFlyby();
            var flock = Birds(sim);
            if (flock.Count >= minSize && flock[0].ColorVariant == 0) return seed;
        }
        return Constants.CANONICAL_TEST_SEED;
    }

    [Fact]
    public void BirdFlybyConstantsArePinnedToSpecValues()
    {
        Assert.Equal(15.0, Constants.BIRD_FLYBY_SPAWN_RATE_PER_HOUR);
        Assert.Equal(7, Constants.BIRD_FLYBY_HOUR_START);
        Assert.Equal(19, Constants.BIRD_FLYBY_HOUR_END);
        Assert.Equal(3, Constants.BIRD_FLOCK_SIZE_MIN);
        Assert.Equal(7, Constants.BIRD_FLOCK_SIZE_MAX);
        Assert.Equal(9.0, Constants.BIRD_FLOCK_FORMATION_SPACING);
        Assert.Equal(22.0, Constants.BIRD_FLOCK_V_ANGLE_DEG);
        Assert.Equal(65.0, Constants.BIRD_SPEED_MIN);
        Assert.Equal(95.0, Constants.BIRD_SPEED_MAX);
        Assert.Equal(78.0, Constants.BIRD_ALTITUDE_MIN);
        Assert.Equal(96.0, Constants.BIRD_ALTITUDE_MAX);
        Assert.Equal(3.6, Constants.BIRD_BODY_LENGTH);
        Assert.Equal(5.0, Constants.BIRD_WING_SPAN);
        Assert.Equal(7.0, Constants.BIRD_WING_FLAP_FREQ);
        Assert.Equal(0.6, Constants.BIRD_WING_FLAP_PHASE_JITTER);
        Assert.Equal(0xFF1A1610u, Constants.BIRD_BODY_COLOR);
        Assert.Equal(1.0, Constants.BIRD_WING_OPEN_RATIO);
        Assert.Equal(0.30, Constants.BIRD_WING_FOLD_RATIO);
        Assert.Equal(0.08, Constants.BIRD_FADE_IN_FRAC);
        Assert.Equal(0.08, Constants.BIRD_FADE_OUT_FRAC);
        Assert.Equal(3.0, Constants.BIRD_DRIFT_AMP_Y);
        Assert.Equal(0.8, Constants.BIRD_DRIFT_FREQ_Y);
        Assert.Equal(0xB12D1F1A1B12D1Aul, Constants.BIRD_FLYBY_PRNG_SALT);
    }

    [Fact]
    public void BirdFlybyPrngSaltIsUnique()
    {
        ulong[] salts =
        {
            Constants.REGROW_PRNG_SALT, Constants.FLOWER_PRNG_SALT, Constants.MUSHROOM_PRNG_SALT,
            Constants.AMBIENT_GUST_PRNG_SALT, Constants.CACTUS_PRNG_SALT, Constants.TUMBLEWEED_PRNG_SALT,
            Constants.CRITTER_PRNG_SALT, Constants.BUTTERFLY_PRNG_SALT, Constants.FIREFLY_PRNG_SALT,
            Constants.SNOWFLAKE_PRNG_SALT, Constants.PINE_PRNG_SALT,
        };
        Assert.All(salts, salt => Assert.NotEqual(Constants.BIRD_FLYBY_PRNG_SALT, salt));
    }

    [Fact]
    public void BirdFlybyFlockSizeStaysInRangeOverSeeds()
    {
        for (ulong i = 0; i < 256; i++)
        {
            ulong seed = unchecked(Constants.CANONICAL_TEST_SEED + i * 0x9E3779B97F4A7C15UL);
            var sim = BuildSim(seed);
            ResetBirdStreamFresh(sim, seed);
            sim.SpawnBirdFlyby();
            Assert.InRange(CountBirds(sim), Constants.BIRD_FLOCK_SIZE_MIN, Constants.BIRD_FLOCK_SIZE_MAX);
        }
    }

    [Fact]
    public void BirdFlybyLeaderAltitudeStaysInRange()
    {
        for (ulong i = 0; i < 128; i++)
        {
            ulong seed = Constants.CANONICAL_TEST_SEED + i;
            var sim = BuildSim(seed);
            ResetBirdStreamFresh(sim, seed);
            sim.SpawnBirdFlyby();
            var flock = Birds(sim);
            Assert.NotEmpty(flock);
            Assert.InRange(flock[0].AltitudeAnchor, Constants.BIRD_ALTITUDE_MIN, Constants.BIRD_ALTITUDE_MAX);
        }
    }

    [Fact]
    public void BirdFlybyLeaderSpeedStaysInRange()
    {
        for (ulong i = 0; i < 128; i++)
        {
            ulong seed = Constants.CANONICAL_TEST_SEED + i;
            var sim = BuildSim(seed);
            ResetBirdStreamFresh(sim, seed);
            sim.SpawnBirdFlyby();
            var flock = Birds(sim);
            Assert.NotEmpty(flock);
            Assert.InRange(flock[0].BaseSpeed, Constants.BIRD_SPEED_MIN, Constants.BIRD_SPEED_MAX);
            Assert.Equal(flock[0].BaseSpeed, Math.Abs(flock[0].Vx), 9);
        }
    }

    [Fact]
    public void BirdFlybyPrngDrawOrderMatchesSideStream()
    {
        const ulong seed = 0xB17D5EED1234UL;
        var sim = BuildSim(seed);
        ResetBirdStreamFresh(sim, seed);
        var side = Prng.Init(seed ^ Constants.BIRD_FLYBY_PRNG_SALT);

        sim.SpawnBirdFlyby();
        int expectedCount = PrngCount(ref side, Constants.BIRD_FLOCK_SIZE_MIN, Constants.BIRD_FLOCK_SIZE_MAX);
        ulong directionBit = side.NextU64() & 1UL;
        double direction = directionBit != 0UL ? 1.0 : -1.0;
        double leaderAltitude = side.Uniform(Constants.BIRD_ALTITUDE_MIN, Constants.BIRD_ALTITUDE_MAX);
        double leaderSpeed = side.Uniform(Constants.BIRD_SPEED_MIN, Constants.BIRD_SPEED_MAX);
        ulong formationStyle = side.NextU64() & 1UL;

        var wingPhases = new List<double>();
        var driftPhases = new List<double>();
        for (int i = 0; i < expectedCount; i++)
        {
            wingPhases.Add(side.Uniform(-Constants.BIRD_WING_FLAP_PHASE_JITTER, Constants.BIRD_WING_FLAP_PHASE_JITTER));
            driftPhases.Add(side.Uniform(0.0, 2.0 * Math.PI));
        }

        var flock = Birds(sim);
        Assert.Equal(expectedCount, flock.Count);
        Assert.Equal(side.State, sim.BirdFlybyPrng.State);

        double spawnX = direction > 0.0 ? -50.0 : Monitor1920 + 50.0;
        double sinAngle = Math.Sin(Constants.BIRD_FLOCK_V_ANGLE_DEG * Math.PI / 180.0);
        for (int i = 0; i < expectedCount; i++)
        {
            double along = -i * Constants.BIRD_FLOCK_FORMATION_SPACING;
            double perpendicular;
            if (formationStyle == 0UL)
            {
                int armIndex = (i + 1) / 2;
                double sideSign = (i % 2) == 0 ? 1.0 : -1.0;
                perpendicular = sideSign * armIndex * Constants.BIRD_FLOCK_FORMATION_SPACING * sinAngle;
            }
            else
            {
                perpendicular = i * Constants.BIRD_FLOCK_FORMATION_SPACING * sinAngle;
            }

            Entity e = flock[i];
            Assert.Equal(spawnX + direction * along, e.X0, 9);
            Assert.Equal(e.X0, e.X, 9);
            Assert.Equal(direction * leaderSpeed, e.Vx, 9);
            Assert.Equal(leaderSpeed, e.BaseSpeed, 9);
            Assert.Equal(leaderAltitude - perpendicular, e.AltitudeAnchor, 9);
            Assert.Equal(wingPhases[i], e.PhaseX, 9);
            Assert.Equal(driftPhases[i], e.PhaseY, 9);
            Assert.Equal(along, e.FormationOffsetAlongFlight, 9);
            Assert.Equal(perpendicular, e.FormationOffsetPerpendicular, 9);
            Assert.Equal((byte)formationStyle, e.ColorVariant);
            Assert.Equal(sim.GlobalTime, e.SpawnTime, 9);
        }
    }

    [Fact]
    public void BirdFlybysAreGrassSceneOnly()
    {
        foreach (Scene scene in new[] { Scene.Desert, Scene.Winter })
        {
            var sim = BuildSim();
            sim.SetScene(scene);
            sim.Entities.Clear();
            ResetBirdSchedule(sim, Constants.CANONICAL_TEST_SEED);
            for (int i = 0; i < 8 * 3600; i++)
            {
                sim.GlobalTime += 1.0;
                sim.TickBirdFlybys(12);
            }
            Assert.Equal(0, CountBirds(sim));
        }
    }

    [Fact]
    public void BirdFlybyDayGatingControlsPoissonSpawns()
    {
        var night = BuildSim();
        ResetBirdSchedule(night, Constants.CANONICAL_TEST_SEED);
        for (int i = 0; i < 10 * 3600; i++)
        {
            night.GlobalTime += 1.0;
            night.TickBirdFlybys(2);
        }
        Assert.Equal(0, CountBirds(night));

        const ulong seed = 0xDAD1B17DUL;
        var day = BuildSim(seed);
        ResetBirdSchedule(day, seed);
        int flybys = 0;
        for (int i = 0; i < 10 * 3600; i++)
        {
            day.GlobalTime += 1.0;
            int before = CountBirds(day);
            day.TickBirdFlybys(12);
            if (CountBirds(day) > before)
            {
                flybys++;
                day.Entities.Clear();
            }
        }

        double observedPerHour = flybys / 10.0;
        Assert.InRange(observedPerHour,
            Constants.BIRD_FLYBY_SPAWN_RATE_PER_HOUR * 0.85,
            Constants.BIRD_FLYBY_SPAWN_RATE_PER_HOUR * 1.15);
    }

    [Fact]
    public void BirdVFormationGeometryIsLocked()
    {
        ulong seed = FindVSeed(5);
        var sim = BuildSim(seed);
        ResetBirdStreamFresh(sim, seed);
        sim.SpawnBirdFlyby();
        var flock = Birds(sim);
        Assert.True(flock.Count >= 5);
        Assert.Equal(0, flock[0].ColorVariant);
        Assert.Equal(0.0, flock[0].FormationOffsetAlongFlight, 9);

        for (int i = 1; i < flock.Count; i++)
        {
            Assert.True(Math.Abs(flock[0].FormationOffsetAlongFlight) < Math.Abs(flock[i].FormationOffsetAlongFlight));
            Assert.Equal(Constants.BIRD_FLOCK_FORMATION_SPACING,
                flock[i - 1].FormationOffsetAlongFlight - flock[i].FormationOffsetAlongFlight, 9);
            double expectedSign = (i % 2 == 0) ? 1.0 : -1.0;
            Assert.Equal(expectedSign, flock[i].FormationOffsetPerpendicular > 0.0 ? 1.0 : -1.0);
        }
    }

    [Fact]
    public void BirdWingFlapScaleStaysInRange()
    {
        for (int i = 0; i < 200; i++)
        {
            double t = i * 0.137;
            double phase = -Constants.BIRD_WING_FLAP_PHASE_JITTER
                + (2.0 * Constants.BIRD_WING_FLAP_PHASE_JITTER) * i / 199.0;
            double scale = Constants.BirdWingScale(t, phase);
            Assert.InRange(scale, Constants.BIRD_WING_FOLD_RATIO, Constants.BIRD_WING_OPEN_RATIO);
        }
    }

    [Fact]
    public void BirdWingPhasesDecorrelateWithinFlock()
    {
        for (ulong i = 1; i < 10000; i++)
        {
            ulong seed = Constants.CANONICAL_TEST_SEED + i;
            var sim = BuildSim(seed);
            ResetBirdStreamFresh(sim, seed);
            sim.SpawnBirdFlyby();
            var flock = Birds(sim);
            if (flock.Count != 5) continue;

            var distinct = new List<double>();
            foreach (Entity e in flock)
            {
                double scale = Constants.BirdWingScale(1.234, e.PhaseX);
                if (!distinct.Any(existing => Math.Abs(existing - scale) < 1e-6))
                {
                    distinct.Add(scale);
                }
            }
            if (distinct.Count >= 3)
            {
                Assert.True(distinct.Count >= 3);
                return;
            }
        }
        Assert.Fail("no decorrelated 5-bird flock found");
    }

    [Fact]
    public void BirdsDespawnPastOppositeBoundary()
    {
        var sim = BuildSim();
        sim.CurrentScene = Scene.Desert;
        sim.Entities.Clear();
        sim.Entities.Add(new Entity
        {
            Kind = EntityKind.Bird,
            X = Monitor1920 + 49.0,
            Y = 10.0,
            Vx = 20.0,
            AltitudeAnchor = Constants.BIRD_ALTITUDE_MIN,
            Lifetime = -1.0,
        });

        sim.TickEntities(0.2);

        Assert.Equal(0, CountBirds(sim));
    }

    [Fact]
    public void BirdsDoNotInteractWithCutsOrCritters()
    {
        var sim = BuildSim();
        sim.Entities.Clear();
        var bird = new Entity
        {
            Kind = EntityKind.Bird,
            X = 500.0,
            Y = sim.WindowHeight - Constants.STRIP_HEIGHT - 10.0,
            Vx = Constants.BIRD_SPEED_MIN,
            BaseSpeed = Constants.BIRD_SPEED_MIN,
            AltitudeAnchor = Constants.BIRD_ALTITUDE_MIN,
            Lifetime = -1.0,
        };
        sim.Entities.Add(bird);
        sim.Entities.Add(new Entity
        {
            Kind = EntityKind.Sheep,
            X = bird.X,
            Y = sim.WindowHeight - Constants.SHEEP_BODY_HEIGHT - Constants.SHEEP_LEG_LENGTH,
            Vx = Constants.SHEEP_WALK_SPEED_MIN,
            State = Constants.SHEEP_STATE_WALKING,
            StateTimer = 10.0,
        });

        sim.ApplyClick(bird.X, bird.Y, 0.0);

        Assert.Equal(EntityKind.Bird, sim.Entities[0].Kind);
        Assert.Equal(Constants.BIRD_SPEED_MIN, sim.Entities[0].BaseSpeed, 9);
        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[1].State);
        Assert.All(sim.Blades, b => Assert.True(b.CutAnimStart < 0.0));
    }

    [Fact]
    public void BirdFlybyPoissonInterArrivalsKeepExpectedMean()
    {
        const ulong seed = 0x510B17D00UL;
        var sim = BuildSim(seed);
        ResetBirdSchedule(sim, seed);

        double prev = sim.GlobalTime;
        double totalInterval = 0.0;
        const int eventsCount = 100;
        for (int i = 0; i < eventsCount; i++)
        {
            sim.GlobalTime = sim.NextBirdFlybyAtTime;
            totalInterval += sim.GlobalTime - prev;
            prev = sim.GlobalTime;
            sim.TickBirdFlybys(12);
            sim.Entities.Clear();
        }

        double expectedMean = 3600.0 / Constants.BIRD_FLYBY_SPAWN_RATE_PER_HOUR;
        double observedMean = totalInterval / eventsCount;
        Assert.InRange(observedMean, expectedMean * 0.80, expectedMean * 1.20);
    }
}

File: tests/DesktopGrass.Native.Tests/src/bird_flyby_tests.cpp
// bird_flyby_tests.cpp - §17.8 ambient bird flyby tests.

#include "TestHelpers.h"
#include "Sim.h"

#include <algorithm>
#include <cmath>
#include <vector>

using namespace desktopgrass;

namespace {
constexpr double Monitor1920 = 1920.0;
constexpr double TwoPi = 6.28318530717958647692;

Sim build_sim(uint64_t seed = CANONICAL_TEST_SEED) {
    Sim sim = sim_init(seed, Monitor1920, DEFAULT_DENSITY);
    sim.currentScene = Scene::Grass;
    sim.entities.clear();
    return sim;
}

int count_birds(const Sim& sim) {
    return static_cast<int>(std::count_if(sim.entities.begin(), sim.entities.end(),
        [](const Entity& e) { return e.kind == EntityKind::Bird; }));
}

std::vector<Entity> birds(const Sim& sim) {
    std::vector<Entity> out;
    for (const Entity& e : sim.entities) {
        if (e.kind == EntityKind::Bird) out.push_back(e);
    }
    return out;
}

int prng_count(Prng& side, int minCount, int maxCount) {
    const double draw = prng_uniform(side, static_cast<double>(minCount), static_cast<double>(maxCount + 1));
    int count = static_cast<int>(std::floor(draw));
    if (count < minCount) count = minCount;
    if (count > maxCount) count = maxCount;
    return count;
}

void reset_bird_stream_fresh(Sim& sim, uint64_t seed) {
    prng_init(sim.birdFlybyPrng, seed ^ BIRD_FLYBY_PRNG_SALT);
    sim.nextBirdFlybyAtTime = sim.globalTime;
}

void reset_bird_schedule(Sim& sim, uint64_t seed) {
    prng_init(sim.birdFlybyPrng, seed ^ BIRD_FLYBY_PRNG_SALT);
    sim.nextBirdFlybyAtTime = sim.globalTime + bird_flyby_sample_interval(sim.birdFlybyPrng);
}

uint64_t find_seed_for_flock_size(int size) {
    for (uint64_t i = 1; i < 10000; ++i) {
        Sim sim = build_sim(CANONICAL_TEST_SEED + i);
        reset_bird_stream_fresh(sim, CANONICAL_TEST_SEED + i);
        sim_spawn_bird_flyby(sim);
        if (count_birds(sim) == size) return CANONICAL_TEST_SEED + i;
    }
    return CANONICAL_TEST_SEED;
}

uint64_t find_v_seed(int minSize) {
    for (uint64_t i = 1; i < 10000; ++i) {
        Sim sim = build_sim(CANONICAL_TEST_SEED + i);
        reset_bird_stream_fresh(sim, CANONICAL_TEST_SEED + i);
        sim_spawn_bird_flyby(sim);
        auto flock = birds(sim);
        if (static_cast<int>(flock.size()) >= minSize && flock[0].colorVariant == 0) {
            return CANONICAL_TEST_SEED + i;
        }
    }
    return CANONICAL_TEST_SEED;
}
} // namespace

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(BirdFlybyTests)
{
public:
TEST_METHOD(BirdFlybyConstantsArePinnedToSpecValues) {
    Assert::IsTrue(BIRD_FLYBY_SPAWN_RATE_PER_HOUR == Near(15.0));
    Assert::IsTrue(BIRD_FLOCK_SIZE_MIN == 3);
    Assert::IsTrue(BIRD_FLOCK_SIZE_MAX == 7);
    Assert::IsTrue(BIRD_FLOCK_FORMATION_SPACING == Near(9.0));
    Assert::IsTrue(BIRD_FLOCK_V_ANGLE_DEG == Near(22.0));
    Assert::IsTrue(BIRD_SPEED_MIN == Near(65.0));
    Assert::IsTrue(BIRD_SPEED_MAX == Near(95.0));
    Assert::IsTrue(BIRD_ALTITUDE_MIN == Near(78.0));
    Assert::IsTrue(BIRD_ALTITUDE_MAX == Near(96.0));
    Assert::IsTrue(BIRD_BODY_LENGTH == Near(3.6));
    Assert::IsTrue(BIRD_WING_SPAN == Near(5.0));
    Assert::IsTrue(BIRD_WING_FLAP_FREQ == Near(7.0));
    Assert::IsTrue(BIRD_WING_FLAP_PHASE_JITTER == Near(0.6));
    Assert::IsTrue(BIRD_BODY_COLOR == 0xFF1A1610u);
    Assert::IsTrue(BIRD_WING_OPEN_RATIO == Near(1.0));
    Assert::IsTrue(BIRD_WING_FOLD_RATIO == Near(0.30));
    Assert::IsTrue(BIRD_FADE_IN_FRAC == Near(0.08));
    Assert::IsTrue(BIRD_FADE_OUT_FRAC == Near(0.08));
    Assert::IsTrue(BIRD_DRIFT_AMP_Y == Near(3.0));
    Assert::IsTrue(BIRD_DRIFT_FREQ_Y == Near(0.8));
    Assert::IsTrue(BIRD_FLYBY_PRNG_SALT == 0xB12D1F1A1B12D1Aull);
}

TEST_METHOD(BirdFlybyPRNGSaltIsUnique) {
    const uint64_t salts[] = {
        REGROW_PRNG_SALT, FLOWER_PRNG_SALT, MUSHROOM_PRNG_SALT,
        AMBIENT_GUST_PRNG_SALT, CACTUS_PRNG_SALT, TUMBLEWEED_PRNG_SALT,
        CRITTER_PRNG_SALT, BUTTERFLY_PRNG_SALT, FIREFLY_PRNG_SALT,
        SNOWFLAKE_PRNG_SALT, PINE_PRNG_SALT,
    };
    for (uint64_t salt : salts) {
        Assert::IsTrue(BIRD_FLYBY_PRNG_SALT != salt);
    }
}

TEST_METHOD(BirdFlybyFlockSizeStaysInRangeOverSeeds) {
    for (uint64_t i = 0; i < 256; ++i) {
        const uint64_t seed = CANONICAL_TEST_SEED + i * 0x9E3779B97F4A7C15ull;
        Sim sim = build_sim(seed);
        reset_bird_stream_fresh(sim, seed);
        sim_spawn_bird_flyby(sim);
        Assert::IsTrue(count_birds(sim) >= BIRD_FLOCK_SIZE_MIN);
        Assert::IsTrue(count_birds(sim) <= BIRD_FLOCK_SIZE_MAX);
    }
}

TEST_METHOD(BirdFlybyLeaderAltitudeStaysInRange) {
    for (uint64_t i = 0; i < 128; ++i) {
        const uint64_t seed = CANONICAL_TEST_SEED + i;
        Sim sim = build_sim(seed);
        reset_bird_stream_fresh(sim, seed);
        sim_spawn_bird_flyby(sim);
        auto flock = birds(sim);
        Assert::IsTrue(!flock.empty());
        Assert::IsTrue(flock[0].altitudeAnchor >= BIRD_ALTITUDE_MIN);
        Assert::IsTrue(flock[0].altitudeAnchor <  BIRD_ALTITUDE_MAX);
    }
}

TEST_METHOD(BirdFlybyLeaderSpeedStaysInRange) {
    for (uint64_t i = 0; i < 128; ++i) {
        const uint64_t seed = CANONICAL_TEST_SEED + i;
        Sim sim = build_sim(seed);
        reset_bird_stream_fresh(sim, seed);
        sim_spawn_bird_flyby(sim);
        auto flock = birds(sim);
        Assert::IsTrue(!flock.empty());
        Assert::IsTrue(flock[0].baseSpeed >= BIRD_SPEED_MIN);
        Assert::IsTrue(flock[0].baseSpeed <  BIRD_SPEED_MAX);
        Assert::IsTrue(std::abs(flock[0].vx) == Near(flock[0].baseSpeed));
    }
}

TEST_METHOD(BirdFlybyPRNGDrawOrderMatchesSideStream) {
    const uint64_t seed = 0xB17D5EED1234ull;
    Sim sim = build_sim(seed);
    reset_bird_stream_fresh(sim, seed);
    Prng side;
    prng_init(side, seed ^ BIRD_FLYBY_PRNG_SALT);

    sim_spawn_bird_flyby(sim);
    const int expectedCount = prng_count(side, BIRD_FLOCK_SIZE_MIN, BIRD_FLOCK_SIZE_MAX);
    const uint64_t directionBit = prng_next_u64(side) & 1ull;
    const double direction = directionBit != 0ull ? 1.0 : -1.0;
    const double leaderAltitude = prng_uniform(side, BIRD_ALTITUDE_MIN, BIRD_ALTITUDE_MAX);
    const double leaderSpeed = prng_uniform(side, BIRD_SPEED_MIN, BIRD_SPEED_MAX);
    const uint64_t formationStyle = prng_next_u64(side) & 1ull;

    std::vector<double> wingPhases;
    std::vector<double> driftPhases;
    for (int i = 0; i < expectedCount; ++i) {
        wingPhases.push_back(prng_uniform(side, -BIRD_WING_FLAP_PHASE_JITTER, BIRD_WING_FLAP_PHASE_JITTER));
        driftPhases.push_back(prng_uniform(side, 0.0, TwoPi));
    }

    auto flock = birds(sim);
    Assert::IsTrue(static_cast<int>(flock.size()) == expectedCount);
    Assert::IsTrue(sim.birdFlybyPrng.state == side.state);

    const double spawnX = direction > 0.0 ? -50.0 : Monitor1920 + 50.0;
    const double sinAngle = std::sin(BIRD_FLOCK_V_ANGLE_DEG * 3.14159265358979323846 / 180.0);
    for (int i = 0; i < expectedCount; ++i) {
        const double along = -static_cast<double>(i) * BIRD_FLOCK_FORMATION_SPACING;
        double perpendicular = 0.0;
        if (formationStyle == 0ull) {
            const int armIndex = (i + 1) / 2;
            const double sideSign = (i % 2) == 0 ? 1.0 : -1.0;
            perpendicular = sideSign * static_cast<double>(armIndex) * BIRD_FLOCK_FORMATION_SPACING * sinAngle;
        } else {
            perpendicular = static_cast<double>(i) * BIRD_FLOCK_FORMATION_SPACING * sinAngle;
        }

        const Entity& e = flock[static_cast<std::size_t>(i)];
        Assert::IsTrue(e.x0 == Near(spawnX + direction * along));
        Assert::IsTrue(e.x == Near(e.x0));
        Assert::IsTrue(e.vx == Near(direction * leaderSpeed));
        Assert::IsTrue(e.baseSpeed == Near(leaderSpeed));
        Assert::IsTrue(e.altitudeAnchor == Near(leaderAltitude - perpendicular));
        Assert::IsTrue(e.phaseX == Near(wingPhases[static_cast<std::size_t>(i)]));
        Assert::IsTrue(e.phaseY == Near(driftPhases[static_cast<std::size_t>(i)]));
        Assert::IsTrue(e.formationOffsetAlongFlight == Near(along));
        Assert::IsTrue(e.formationOffsetPerpendicular == Near(perpendicular));
        Assert::IsTrue(e.colorVariant == static_cast<uint8_t>(formationStyle));
        Assert::IsTrue(e.spawnTime == Near(sim.globalTime));
    }
}

TEST_METHOD(BirdFlybysAreGrassSceneOnly) {
    for (Scene scene : { Scene::Desert, Scene::Winter }) {
        Sim sim = build_sim();
        sim_set_scene(sim, scene);
        sim.entities.clear();
        reset_bird_schedule(sim, CANONICAL_TEST_SEED);
        for (int i = 0; i < 8 * 3600; ++i) {
            sim.globalTime += 1.0;
            sim_tick_bird_flybys(sim);
        }
        Assert::IsTrue(count_birds(sim) == 0);
    }
}

TEST_METHOD(BirdFlybyPoissonSpawnsWhenScheduleElapses) {
    Sim sim = build_sim(0xDAD1B17Dull);
    reset_bird_schedule(sim, 0xDAD1B17Dull);
    int flybys = 0;
    for (int i = 0; i < 10 * 3600; ++i) {
        sim.globalTime += 1.0;
        const int before = count_birds(sim);
        sim_tick_bird_flybys(sim);
        if (count_birds(sim) > before) {
            ++flybys;
            sim.entities.clear();
        }
    }

    const double observedPerHour = static_cast<double>(flybys) / 10.0;
    Assert::IsTrue(observedPerHour == Near(BIRD_FLYBY_SPAWN_RATE_PER_HOUR).epsilon(0.15));
}

TEST_METHOD(BirdVFormationGeometryIsLocked) {
    const uint64_t seed = find_v_seed(5);
    Sim sim = build_sim(seed);
    reset_bird_stream_fresh(sim, seed);
    sim_spawn_bird_flyby(sim);
    auto flock = birds(sim);
    Assert::IsTrue(flock.size() >= 5);
    Assert::IsTrue(flock[0].colorVariant == 0);
    Assert::IsTrue(flock[0].formationOffsetAlongFlight == Near(0.0));

    for (std::size_t i = 1; i < flock.size(); ++i) {
        Assert::IsTrue(std::fabs(flock[0].formationOffsetAlongFlight)
            < std::fabs(flock[i].formationOffsetAlongFlight));
        Assert::IsTrue(flock[i - 1].formationOffsetAlongFlight - flock[i].formationOffsetAlongFlight
            == Near(BIRD_FLOCK_FORMATION_SPACING));
        const double expectedSign = (i % 2 == 0) ? 1.0 : -1.0;
        Assert::IsTrue((flock[i].formationOffsetPerpendicular > 0.0 ? 1.0 : -1.0) == expectedSign);
    }
}

TEST_METHOD(BirdWingFlapScaleStaysInRange) {
    for (int i = 0; i < 200; ++i) {
        const double t = i * 0.137;
        const double phase = -BIRD_WING_FLAP_PHASE_JITTER
            + (2.0 * BIRD_WING_FLAP_PHASE_JITTER) * (static_cast<double>(i) / 199.0);
        const double scale = bird_wing_scale(t, phase);
        Assert::IsTrue(scale >= BIRD_WING_FOLD_RATIO);
        Assert::IsTrue(scale <= BIRD_WING_OPEN_RATIO);
    }
}

TEST_METHOD(BirdWingPhasesDecorrelateWithinAFlock) {
    for (uint64_t i = 1; i < 10000; ++i) {
        const uint64_t seed = CANONICAL_TEST_SEED + i;
        Sim sim = build_sim(seed);
        reset_bird_stream_fresh(sim, seed);
        sim_spawn_bird_flyby(sim);
        auto flock = birds(sim);
        if (flock.size() != 5) continue;

        std::vector<double> distinct;
        for (const Entity& e : flock) {
            const double scale = bird_wing_scale(1.234, e.phaseX);
            bool seen = false;
            for (double existing : distinct) {
                if (std::fabs(existing - scale) < 1e-6) { seen = true; break; }
            }
            if (!seen) distinct.push_back(scale);
        }
        if (distinct.size() >= 3) {
            Assert::IsTrue(distinct.size() >= 3);
            return;
        }
    }
    Assert::Fail(L"No decorrelated 5-bird flock found", LINE_INFO());
}

TEST_METHOD(BirdsDespawnPastOppositeBoundary) {
    Sim sim = build_sim();
    sim.currentScene = Scene::Desert;
    sim.entities.clear();
    Entity bird{};
    bird.kind = EntityKind::Bird;
    bird.x = Monitor1920 + 49.0;
    bird.y = 10.0;
    bird.vx = 20.0;
    bird.altitudeAnchor = BIRD_ALTITUDE_MIN;
    bird.lifetime = -1.0;
    sim.entities.push_back(bird);

    sim_tick_entities(sim, 0.2);

    Assert::IsTrue(count_birds(sim) == 0);
}

TEST_METHOD(BirdsDoNotInteractWithCutsOrCritters) {
    Sim sim = build_sim();
    sim.entities.clear();
    Entity bird{};
    bird.kind = EntityKind::Bird;
    bird.x = 500.0;
    bird.y = sim.windowHeight - STRIP_HEIGHT - 10.0;
    bird.vx = BIRD_SPEED_MIN;
    bird.baseSpeed = BIRD_SPEED_MIN;
    bird.altitudeAnchor = BIRD_ALTITUDE_MIN;
    bird.lifetime = -1.0;
    sim.entities.push_back(bird);

    Entity sheep{};
    sheep.kind = EntityKind::Sheep;
    sheep.x = bird.x;
    sheep.y = sim.windowHeight - SHEEP_BODY_HEIGHT - SHEEP_LEG_LENGTH;
    sheep.vx = SHEEP_WALK_SPEED_MIN;
    sheep.state = SHEEP_STATE_WALKING;
    sheep.stateTimer = 10.0;
    sim.entities.push_back(sheep);

    InputEvent ev{};
    ev.type = EventType::Click;
    ev.x = bird.x;
    ev.y = bird.y;
    sim_apply_click(sim, ev);

    Assert::IsTrue(sim.entities[0].kind == EntityKind::Bird);
    Assert::IsTrue(sim.entities[0].baseSpeed == Near(BIRD_SPEED_MIN));
    Assert::IsTrue(sim.entities[1].state == SHEEP_STATE_WALKING);
    for (const Blade& b : sim.blades) Assert::IsTrue(b.cutAnimStart < 0.0);
}

TEST_METHOD(BirdFlybyPoissonInterArrivalsKeepExpectedMean) {
    const uint64_t seed = 0x510B17D00ull;
    Sim sim = build_sim(seed);
    reset_bird_schedule(sim, seed);

    double prev = sim.globalTime;
    double totalInterval = 0.0;
    constexpr int Events = 100;
    for (int i = 0; i < Events; ++i) {
        sim.globalTime = sim.nextBirdFlybyAtTime;
        totalInterval += sim.globalTime - prev;
        prev = sim.globalTime;
        sim_tick_bird_flybys(sim);
        sim.entities.clear();
    }

    const double expectedMean = 3600.0 / BIRD_FLYBY_SPAWN_RATE_PER_HOUR;
    Assert::IsTrue((totalInterval / Events) == Near(expectedMean).epsilon(0.20));
}
};
}
